agora/Agora/Governor.hs

```diff
@@ -1,62 +1,638 @@
+-}
 governorValidator :: Governor -> ClosedTerm PValidator
-governorValidator _ =
-  plam $ \_datum _redeemer _ctx' -> P.do
-    popaque (pconstant ())
+governorValidator gov =
+  plam $ \datum' redeemer' ctx' -> P.do
+    -- TODO: use ptryFrom
+    redeemer <- pmatch $ pfromData @PGovernorRedeemer $ punsafeCoerce redeemer'
+    ctx <- pletFields @'["txInfo", "purpose"] ctx'
+
+    txInfo' <- plet $ pfromData $ ctx.txInfo
+    txInfo <- pletFields @'["mint", "inputs", "outputs"] txInfo'
+
+    PSpending ((pfield @"_0" #) -> txOutRef') <- pmatch $ pfromData ctx.purpose
+    let txOutRef = pfromData txOutRef'
+
+    PJust ((pfield @"resolved" #) -> ownInput') <- pmatch $ pfindTxInByTxOutRef # txOutRef # txInfo'
+    ownInput <- pletFields @'["address", "value"] ownInput'
+    let selfAddress = pfromData $ ownInput.address
+
+    -- TODO: use ptryFrom
+    let oldParams' = pfromData @PGovernorDatum $ punsafeCoerce datum'
+    oldParams <- pletFields @'["proposalThresholds", "nextProposalId"] oldParams'
+
+    let ownInputDatumNFTAmount = stateTokenValueOf # ownInput.value
+    passert "Own input should have exactly one state token" $
+      ownInputDatumNFTAmount #== 1
+
+    ownOutputs <- plet $ findOutputsToAddress # txInfo' # selfAddress
+    passert "Exactly one utxo should be sent to the governor" $
+      plength # ownOutputs #== 1
+
+    ownOutput <- pletFields @'["value", "datumHash"] $ phead # ownOutputs
+    let ownOuputDatumNFTAmount = stateTokenValueOf # ownOutput.value
+    passert "State token should stay at governor's address" $
+      ownOuputDatumNFTAmount #== 1
+    passert "Output utxo to governor should have datum" $
+      pisDJust # ownOutput.datumHash
+
+    let outputGovernorStateDatumHash = mustBePDJust # "Output governor state datum hash not found" # ownOutput.datumHash
+
+    newDatumData <-
+      plet $
+        pforgetData $
+          pdata $
+            mustBePJust # "Ouput governor state datum not found"
+              #$ pfindDatum # outputGovernorStateDatumHash # txInfo'
+
+    case redeemer of
+      PCreateProposal _ -> P.do
+        let expectedNextProposalId = pnextProposalId # oldParams.nextProposalId
+            expectedNewDatum =
+              pcon $
+                PGovernorDatum $
+                  pdcons @"proposalThresholds" # oldParams.proposalThresholds
+                    #$ pdcons @"nextProposalId" # pdata expectedNextProposalId # pdnil
+
+        passert "Unexpected governor state datum" $
+          newDatumData #== pforgetData (pdata expectedNewDatum)
+
+        passert "Exactly one proposal token must be minted" $
+          hasOnlyOneTokenOfCurrencySymbol # pproposalSymbol # txInfo.mint
+
+        outputs <- plet $ findOutputsToAddress # ctx.txInfo # pproposalValidatorAddress
+        passert "Exactly one utxo should be sent to the proposal validator" $
+          plength # outputs #== 1
+
+        output <- pletFields @'["value", "datumHash"] $ phead # outputs
```

Comment:
So we assert that the first output is to the proposal? That seems a little suspicious, besides, we don't even document that anywhere. Could we rewrite this in terms of a fold that scans for the right address? (we know the address of proposal validator within this script!)

Reply:
We _are_ scanning for the proposal validator's address:
```haskell
outputs <- plet $ findOutputsToAddress # ctx.txInfo # pproposalValidatorAddress
```
You meat we scan for the proposal state token that get sent to the right address, right?

Reply:
I can do something like this:
```haskell
filteredOutputs <- plet $ pfilter # plam (\txOut' -> P.do
                                        txOut <- pletFields @'["address", "value"] txOut'

                                        txOut.address #== pdata pproposalValidatorAddress #&&
                                          psymbolValueOf # pproposalSymbol # txOut.value #== 1
                                        ) # pfromData txInfo.outputs

        passert "Exactly one utxo with proposal state token should be sent to the proposal validator" $
          plength # filteredOutputs #== 1
```

Reply:
In other word, this will allow multiple UTXOs to be sent to the proposal validator in a single transaction. I don't know if this is intended or not.